### PR TITLE
121_BestTimeToBuyAndSellStock.cpp

### PR DESCRIPTION
**Pull request by andriibugaiov**

### CODE REVIEW

File: cpp/121_BestTimeToBuyAndSellStock.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    const int BOUGHT_BEFORE_TODAY = 1;
+    const int BOUGHT_BEFORE_TODAY_OR_TODAY = 1;
+    const int INITIAL_BUDGET = 0;
+public:
+    int maxProfit(vector<int>& prices) {
+        // Focus on the end of some trade day (11:59PM).
+        // The two states possible are:
+        // I. We sold stock before today or today (implies that we bought stock before that day).
+        // II. (12:00PM - midnight) We bought stock before today or today.
+        // III. We haven't bought stock yet.
```
Comment (Yorkio) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: Do not get why we need to divide day 12:00PM - midnight in second case.

File: cpp/121_BestTimeToBuyAndSellStock.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            }
+            
+            // Maximize budget for state II.
+            budget_by_state[BOUGHT_BEFORE_TODAY_OR_TODAY] = max(
+                budget_by_state[BOUGHT_BEFORE_TODAY],
+                INITIAL_BUDGET - price_today);
+        }
+        // If the budget is less than the initial budget in the end,
+        // we better to not make a transaction.
+        return max(budget_by_state[SOLD_BEFORE_TODAY_OR_TODAY], INITIAL_BUDGET);
```
Comment (Yorkio) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: Solution seems for me too complicated for this problem. I see that you want to show that it in future can be extended to harder DP problems, but why not just use simpler greedy solution?

File: cpp/121_BestTimeToBuyAndSellStock.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                // Maximize budget for state I.
+                budget_by_state[SOLD_BEFORE_TODAY_OR_TODAY] = max(
+                    budget_by_state[SOLD_BEFORE_TODAY],
+                    budget_by_state[BOUGHT_BEFORE_TODAY] + price_today);
+            }
+            
+            // Maximize budget for state II.
+            budget_by_state[BOUGHT_BEFORE_TODAY_OR_TODAY] = max(
+                budget_by_state[BOUGHT_BEFORE_TODAY],
+                INITIAL_BUDGET - price_today);
```
Comment (Yorkio) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: It is also worth mentioning that in fact by maximizing budget for state II you in fact try to minimize spendings on stocks.

File: cpp/121_BestTimeToBuyAndSellStock.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        vector<int> budget_by_state(2);
+        budget_by_state[SOLD_BEFORE_TODAY_OR_TODAY] = INT_MIN;
+        budget_by_state[BOUGHT_BEFORE_TODAY_OR_TODAY] = INT_MIN;
+        
+        for (int price_today : prices) {
+            // Make sure that we have already bought stock.
+            if (budget_by_state[BOUGHT_BEFORE_TODAY] != INT_MIN) {
+                // Maximize budget for state I.
+                budget_by_state[SOLD_BEFORE_TODAY_OR_TODAY] = max(
+                    budget_by_state[SOLD_BEFORE_TODAY],
```
Comment (Yorkio) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: I am not sure having different names for the same states is good idea.
The meaning of this state is SOLD_BEFORE_TODAY_OR_TODAY in your DP.
I see that before update this state shows only before today, but still it seems redundant.

File: cpp/121_BestTimeToBuyAndSellStock.cpp (hunk 2 above)
Comment (andriibugaiov) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: I just wanted to practice. I guess you are right that I shouldnt have rush with complex solutions. This is why we are solving easy problems now...to practice easy solutions.

File: cpp/121_BestTimeToBuyAndSellStock.cpp (hunk 1 above)
Comment (andriibugaiov) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: Nevermind. I mean that 12pm is a midnight. But nevermind. I have updated the comments.

File: cpp/121_BestTimeToBuyAndSellStock.cpp (hunk 4 above)
Comment (andriibugaiov) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: The sates are different it is that we store them in the same place (same index -> same vector slot).

When you maximize budget when, for example, you sell stock, you read the formula like: the max budget after selling the stock before-today-or-today (one state) is the max of budget after selling the stock before-today (another state) and after selling the stock today. Just try to read the code :)

File: cpp/121_BestTimeToBuyAndSellStock.cpp (hunk 3 above)
Comment (andriibugaiov) [inline cpp/121_BestTimeToBuyAndSellStock.cpp]: Ok.

**Pull request merged**